rllib/algorithms/dt/segmentation_buffer.py

```diff
@@ -0,0 +1,215 @@
+                f"or random shuffling that caused a duplicate rollout (no need to fix)."
+            )
+            episode = episode[: self.max_ep_len]
+
+        if len(self._buffer) < self.capacity:
+            self._buffer.append(episode)
+        else:
+            # TODO: replace proportional to episode length
+            replace_ind = np.random.randint(0, self.capacity)
+            self._buffer[replace_ind] = episode
+
+    def sample(self, batch_size: int) -> SampleBatch:
+        """Sample segments from the buffer.
+
+        Args:
+            batch_size: number of segments to sample.
+
+        Returns:
+            SampleBatch of segments with keys and shape {
+                OBS: [batch_size, max_seq_len, obs_dim],
+                ACTIONS: [batch_size, max_seq_len, act_dim],
+                RETURNS_TO_GO: [batch_size, max_seq_len + 1, 1],
+                T: [batch_size, max_seq_len],
+                ATTENTION_MASKS: [batch_size, max_seq_len],
+            }
+        """
+        samples = [self._sample_single() for _ in range(batch_size)]
+        return concat_samples(samples)
+
+    def _sample_single(self) -> SampleBatch:
+        # TODO: sample proportional to episode length
+        # Sample a random episode from the buffer and then sample a random
+        # segment from that episode.
+        buffer_ind = np.random.randint(0, len(self._buffer))
+        episode = self._buffer[buffer_ind]
+        ep_len = episode[SampleBatch.OBS].shape[0]
+        # This offset accounts for either the normal case when max_seq_len is shorter
+        # than the episode length, but also occasionally an episode will be shorter
+        # than the context length (max_seq_len).
+        offset = min(self.max_seq_len, ep_len)
+        # We allow si to be negative (for now) because we want segments that only
+        # contains the first few transitions (and padd the rest),
+        # for example [0, 0, 0, 0, 0, 0, R0, s0, a0].
+        si = np.random.randint(-offset + 1, ep_len - offset + 1)
+        ei = si + offset
+        # but for actual segmenting we don't want starting index to be negative.
+        si = max(si, 0)
+
+        obs = episode[SampleBatch.OBS][si:ei]
+        actions = episode[SampleBatch.ACTIONS][si:ei]
+        # Note that returns-to-go needs an extra elem as the target for the last action.
+        returns_to_go = episode[SampleBatch.RETURNS_TO_GO][si : ei + 1].reshape(-1, 1)
+
+        length = obs.shape[0]
+        timesteps = np.arange(si, si + length, dtype=np.int32)
+        masks = np.ones(length, dtype=returns_to_go.dtype)
+
+        # Back pad returns-to-go with 0 if at end of rollout.
+        if returns_to_go.shape[0] == length:
+            returns_to_go = np.concatenate(
+                [returns_to_go, np.zeros((1, 1), dtype=returns_to_go.dtype)], axis=0
+            )
+
+        # Front-pad if at beginning of rollout.
+        pad_length = self.max_seq_len - length
+        if pad_length > 0:
```

Comment:
**TODO**: I want you think about simplification of this padding logic, and re-use general purpose code as much as you can. For example, define a `pad_left_with_zeros(arr, max_ep_len)` function that left pads a given arr if the number of elements is less than max_ep_len. it can be used for all entries (including the mask element). Remember that the overall dtype should be the same as the arr.dtype.

Reply:
added